### PR TITLE
chore: encapsulate sparse API

### PR DESCRIPTION
**Pull request by winston-h-zhang**
Small PR to clean up the sparse matrix API, so that manipulating the rows is clear.

### CODE REVIEW

File: src/r1cs/sparse.rs [hunk 1]
```diff
@@ -111,6 +117,30 @@ impl<F: PrimeField> SparseMatrix<F> {
...
+
+  /// Returns a parallel iterator into the rows
+  pub fn par_iter_rows(&self) -> impl IndexedParallelIterator<Item = &RowData> {
+    self
+      .indptr
+      .par_windows(2)
+      .map(|ptrs| RowData::ref_cast(ptrs.try_into().unwrap()))
+  }
+
+  /// Retrieves the data for row slice [i..j] from `row`.
```
Comment (huitseeker) [inline src/r1cs/sparse.rs:136]: It would be good to document that ths returned usize is a column index.

File: src/r1cs/sparse.rs [hunk 2]
```diff
@@ -31,6 +32,11 @@ pub struct SparseMatrix<F: PrimeField> {
   pub cols: usize,
 }
 
+/// Wrapper type for encode rows of [`SparseMatrix`]
+#[derive(Debug, Clone, RefCast)]
+#[repr(transparent)]
+pub struct RowData([usize; 2]);
```
Comment (huitseeker) [inline src/r1cs/sparse.rs:38]: This isn't really `RowData`, is it, rather `RowIndexes`, right?

**Review by huitseeker — APPROVED** [review]
Forgot to approve, the comments are non-blocking.

**Comment by huitseeker** [comment]
!benchmark --bench recursive-snark --features cuda

**Comment by samuelburnham** [comment]
!benchmark --bench recursive-snark --features cuda

**Pull request merged**